### PR TITLE
Select projects and clients from edit view

### LINKED ISSUE #445: Edit view project selection and auto completion
I hope this can reuse the completion from the creation dialog.

It should split the entered string into words and match those against client, project, and task names.

Of course, the user should then be able to choose a project, and upon confirmation, the change has to be persisted.

### PR DESCRIPTION
**Pull request by arturs105**
Ref #445 

This PR adds the functionality to select and filter projects and clients and also return them to the edit time entry view. It does not, however, persist those changes after pressing "Confirm changes".

Do a functional test before approving.

### CODE REVIEW

File: Toggl.Daneel/Combiners/ProjectTaskClientValueCombiner.cs [hunk 1]
```diff
@@ -6,10 +6,7 @@
 using System.Text;
 using UIKit;
 using Foundation;
-using System.Globalization;
```
Comment (willsb) [inline Toggl.Daneel/Combiners/ProjectTaskClientValueCombiner.cs]: Unrelated

File: Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            => tableView.DequeueReusableCell(cellIdentifier, indexPath);
+        
+        public override nfloat GetHeightForRow(UITableView tableView, NSIndexPath indexPath) => 48;
+
+        public override nfloat GetHeightForHeader(UITableView tableView, nint section) => 40;
+
+        public override void HeaderViewDisplayingEnded(UITableView tableView, UIView headerView, nint section)
+        {
+            var firstVisible = TableView.IndexPathsForVisibleRows.First();
+            if (firstVisible.Section == section + 1)
```
Comment (willsb) [inline Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs]: Early return here instead

File: Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+        public override void WillDisplayHeaderView(UITableView tableView, UIView headerView, nint section)
+        {
+            var firstVisibleIndexPath = TableView.IndexPathsForVisibleRows.First();
+            if (firstVisibleIndexPath.Section == section)
+            {
+                var nextHeader = (SelectProjectHeaderViewCell)TableView.GetHeaderView(section + 1);
+                if (nextHeader == null) return;
+                nextHeader.TopSeparatorView.Hidden = false;
+                ((SelectProjectHeaderViewCell)headerView).TopSeparatorView.Hidden = true;
```
Comment (willsb) [inline Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs]: Maybe cast this at the top, assign `TopSeparatorView.Hidden` and early return if needed?

File: Toggl.Daneel/Views/SelectProjectHeaderViewCell.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+using UIKit;
+
+namespace Toggl.Daneel.Views
+{
+    public sealed partial class SelectProjectHeaderViewCell : MvxTableViewHeaderFooterView
+    {
+        public static readonly NSString Key = new NSString(nameof(SelectProjectHeaderViewCell));
+        public static readonly UINib Nib;
+
+        public UIView TopSeparatorView => TopSeparator;
```
Comment (willsb) [inline Toggl.Daneel/Views/SelectProjectHeaderViewCell.cs]: Is `TopSeparator` protected?

File: Toggl.Foundation.MvvmCross/ViewModels/ProjectSuggestionCollection.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        public string Workspace { get; }
+
+        public ProjectSuggestionCollection(string workspace, IEnumerable<ProjectSuggestion> suggestions)
+        {
+            Ensure.Argument.IsNotNull(workspace, nameof(workspace));
+            Ensure.Argument.IsNotNull(suggestions, nameof(suggestions));
+
+            Workspace = workspace;
+            Add(new ProjectSuggestion());
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/ViewModels/ProjectSuggestionCollection.cs]: I don't like this to be made public. There's no need to instantiate multiple objects; expose this as `ProjectSuggestion.NoProject` instead

Comment (willsb) [reply Toggl.Foundation.MvvmCross/ViewModels/ProjectSuggestionCollection.cs]: Also, do we really need the `No Project` suggestion to appear on every workspace?

File: Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs [hunk 6]
```diff
@@ -164,6 +170,31 @@ private Task close()
...
+            var selectedProjectId = await navigationService.Navigate<SelectProjectViewModel, IdParameter, IdParameter>(parameter);
+
+            if (selectedProjectId.Id == projectId)
+                return;
+
+            projectId = selectedProjectId.Id;
+
+            if (projectId == 0)
+            {
+                Project = "";
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs]: Make all these assignments inline

File: Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs [hunk 7]
```diff
@@ -164,6 +170,31 @@ private Task close()
             StartTime = selectedDate.GetDate();
         }
 
+        private async Task selectProject()
+        {
+            var parameter = IdParameter.WithId(projectId);
+            var selectedProjectId = await navigationService.Navigate<SelectProjectViewModel, IdParameter, IdParameter>(parameter);
+
+            if (selectedProjectId.Id == projectId)
+                return;
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs]: No need for a newline here

File: Toggl.Foundation.MvvmCross/ViewModels/SelectProjectViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+            OnTextChanged();
+        }
+
+        private void OnTextChanged()
+        {
+            infoSubject.OnNext(Text);
+        }
+
+        private void onSuggestions(IEnumerable<AutocompleteSuggestion> suggestions)
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/ViewModels/SelectProjectViewModel.cs]: Since you can only ever have one type of suggestions, why not cast this on the observable declaration?

File: Toggl.Foundation.MvvmCross/ViewModels/SelectProjectViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+        public override void Prepare(IdParameter parameter)
+            => projectId = parameter.Id;
+
+        public override async Task Initialize()
+        {
+            await base.Initialize();
+
+            infoSubject.AsObservable()
+                .SelectMany(t => dataSource.AutocompleteProvider.Query(t, AutocompleteSuggestionType.Projects))
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/ViewModels/SelectProjectViewModel.cs]: Give a meaningful name to this `t`

File: Toggl.Foundation.MvvmCross/ViewModels/SelectProjectViewModel.cs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+                .Subscribe(onSuggestions);
+
+            OnTextChanged();
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/ViewModels/SelectProjectViewModel.cs]: Use `dataSource.AutocompleteProvider.Query(t, AutocompleteSuggestionType.Projects).StartWith()` or something like that to combine observables and making your intentions clearer instead of calling this method here

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/SelectProjectViewModelTests.cs [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+            }
+
+            [Fact]
+            public void ReturnsTheSelectedProjectId()
+            {
+                var selectedProject = new ProjectSuggestion();
+                ViewModel.SelectProjectCommand.ExecuteAsync(selectedProject).Wait();
+
+                NavigationService.Received()
+                    .Close(Arg.Is(ViewModel), Arg.Is<IdParameter>(p => p.Id == selectedProject.ProjectId));
```
Comment (willsb) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/SelectProjectViewModelTests.cs]: This test is kinda bad because you're not initializing the VM with a different project id and the default value for long is 0 (the same value an empty `ProjectSuggestion` returns

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/SelectProjectViewModelTests.cs [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+            }
+        }
+
+        public sealed class TheTextProperty : SelectProjectViewModelTest
+        {
+            [Fact]
+            public async Task WhenChangedTheAutocompleteProviderIsQueried()
```
Comment (willsb) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/SelectProjectViewModelTests.cs]: `WhenChangedQueriesTheAutocompleteProvider`

File: Toggl.Foundation/Autocomplete/Suggestions/ProjectSuggestion.cs [hunk 13]
```diff
@@ -44,9 +46,11 @@ public ProjectSuggestion(IDatabaseProject project)
             ProjectName = project.Name;
             ProjectColor = project.Color;
 
-            if (project.Client == null) return;
-
-            ClientName = project.Client.Name;
+            if (project.Workspace != null)
+                Workspace = project.Workspace.Name;
```
Comment (willsb) [inline Toggl.Foundation/Autocomplete/Suggestions/ProjectSuggestion.cs]: Since you're only accessing Workspace once, why not use the elvis operator? (similar for Client)

File: Toggl.Daneel/Views/SelectProjectHeaderViewCell.cs (hunk 4 above)
Comment (arturs105) [reply Toggl.Daneel/Views/SelectProjectHeaderViewCell.cs]: It has no access modifier in the `SelectProjectHeaderViewCell.designer.cs`, so i think it's private.

File: Toggl.Foundation.MvvmCross/ViewModels/ProjectSuggestionCollection.cs (hunk 5 above)
Comment (arturs105) [reply Toggl.Foundation.MvvmCross/ViewModels/ProjectSuggestionCollection.cs]: Yes, Heidi confirmed, that the empty project should be in every workspace.

File: Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs (hunk 3 above)
Comment (arturs105) [reply Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs]: I agree about moving the cast to the top, but i don't see how i can early return, since the both parts of the `if` statement are needed.

File: Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs (hunk 6 above)
Comment (arturs105) [reply Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs]: Do you mean something like this?

```
if (projectId == 0)
{
    Project = ""; Task = ""; Client = ""; ProjectColor = ""; return;
}
```

If so, then i disagree. After seeing this, i had the urge to remove the braces, since it looks like a single statement.

File: Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+            if (nextHeader == null) return;
+
+            nextHeader.TopSeparatorView.Hidden = true;
+        }
+
+        public override void WillDisplayHeaderView(UITableView tableView, UIView headerView, nint section)
+        {
+            var selectProjectHeaderViewCell = (SelectProjectHeaderViewCell)headerView;
+            var firstVisibleIndexPath = TableView.IndexPathsForVisibleRows.First();
+            if (firstVisibleIndexPath.Section == section)
```
Comment (willsb) [inline Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs]:
```
var topIsHidden = firstVisibleIndexPath.Section == section;
selectProjectHeaderViewCell.TopSeparatorView.Hidden = topIsHidden;
if (!topIsHidden) return;
```

File: Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs (hunk 6 above)
Comment (willsb) [reply Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs]:
```csharp
if (projectId == 0)
{
    Project = Task = Client = ProjectColor = ""; 
    return;
}
```

File: Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs [hunk 15]
```diff
@@ -162,6 +168,29 @@ private Task close()
...
 
+        private async Task selectProject()
+        {
+            var selectedProjectId = await navigationService.Navigate<SelectProjectViewModel, long, long>(projectId);
+
+            if (selectedProjectId == projectId) return;
+
+            projectId = selectedProjectId;
+
+            if (projectId == 0)
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs]:
```csharp
if (projectId == 0)
{
    Project = Task = Client = ProjectColor = ""; 
    return;
}
```

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/SelectProjectViewModelTests.cs [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+            {
+                await ViewModel.SelectProjectCommand.ExecuteAsync(ProjectSuggestion.NoProject);
+
+                await NavigationService.Received().Close(Arg.Is(ViewModel), Arg.Any<long>());
+            }
+
+            [Fact]
+            public void ReturnsTheSelectedProjectId()
+            {
+                var selectedProject = ProjectSuggestion.NoProject;
```
Comment (willsb) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/SelectProjectViewModelTests.cs]: Like I said: Please do initialize this test with a value other than 0 so we can see that the selected value is indeed pumped

Comment (willsb) [reply Toggl.Foundation.Tests/MvvmCross/ViewModels/SelectProjectViewModelTests.cs]: Or maybe pass a different selected project here, whichever works best

File: Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs (hunk 15 above)
Comment (arturs105) [reply Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs]: Yes, hadn't used this for a long time, so had completely forgotten about it.

**Comment by arturs105** [comment]
The cursor color was changed [here](https://github.com/toggl/mobileapp/pull/594/commits/3cb464aac2af70bb0549c700c6278417a2317cef) by setting the tint color of the text field.

File: Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs (hunk 14 above)
Comment (arturs105) [reply Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs]: So the whole method looks something like this?

```
var selectProjectHeaderViewCell = (SelectProjectHeaderViewCell)headerView;
var firstVisibleIndexPath = TableView.IndexPathsForVisibleRows.First();
var appearingFromTop = firstVisibleIndexPath.Section == section;

selectProjectHeaderViewCell.TopSeparatorView.Hidden = appearingFromTop;

if (!appearingFromTop) return;

var nextHeader = (SelectProjectHeaderViewCell)TableView.GetHeaderView(section + 1);
if (nextHeader == null) return;
nextHeader.TopSeparatorView.Hidden = false;
```

Comment (willsb) [reply Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs]: Exactly

Comment (willsb) [reply Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs]: Actually, nevermind, leave it as is
